src/Security/AuthenticationMiddleware.php

```diff
@@ -46,9 +103,27 @@ public function setAuthenticationHandler(AuthenticationHandler $authenticationHa
     public function process(HTTPRequest $request, callable $delegate)
     {
         try {
-            $this
-                ->getAuthenticationHandler()
-                ->authenticateRequest($request);
+            $handler = $this->getAuthenticationHandler();
+            $handler->authenticateRequest($request);
+
+            if (static::config()->get('performExpiredPasswordCheck')) {
+                $session = $request->getSession();
+                if ($session && $session->get(static::SESSION_KEY_EXPIRED_PASSWORD_ALLOW_REQUEST)) {
+                    // allow current request and skip the expiration check, but for only the current
+                    // request, so we're deleting the flag from the session so it's not affecting other
+                    // requests.
+                    // This flag would usually be set from within $handler->authenticateRequest()
+
+                    $session->clear(static::SESSION_KEY_EXPIRED_PASSWORD_ALLOW_REQUEST);
+                } else {
+                    $user = Security::getCurrentUser();
+                    if ($user && $user->isPasswordExpired()) {
+                        if ($response = $this->handleExpiredPassword($request)) {
+                            return $response;
+                        }
+                    }
+                }
+            }
```

Comment:
This is bordering on personal opinion, but I think this code would look a lot cleaner being rewritten for early returns:

```php
if (!static::config()->get('performExpiredPasswordCheck')) {
    return $delegate($request);
}

$session = $request->getSession();
if ($session && $session->get(static::SESSION_KEY_EXPIRED_PASSWORD_ALLOW_REQUEST)) {
    $session->clear(static::SESSION_KEY_EXPIRED_PASSWORD_ALLOW_REQUEST);
    return $delegate($request);
}

// ...
```

Reply:
done